rust/benchmark-datasets/src/types.rs

```diff
@@ -0,0 +1,235 @@
+use anyhow::Result;
+use rand::seq::SliceRandom;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, future::Future, path::PathBuf};
+use tantivy::{
+    collector::TopDocs,
+    doc,
+    query::QueryParser,
+    schema::{Schema, STORED, TEXT},
+    Index, ReloadPolicy,
+};
+use tokio::io::AsyncWriteExt;
+use tokio_stream::{Stream, StreamExt};
+
+use crate::util::{get_dir_for_persistent_dataset_files, get_or_populate_cached_dataset_file};
+
+trait TantivyIndexable {
+    fn build_tantivy_index(&mut self) -> impl Future<Output = Result<(Index, Schema)>>;
+}
+
+impl<T: Stream<Item = Document> + Unpin> TantivyIndexable for T {
+    async fn build_tantivy_index(&mut self) -> Result<(Index, Schema)> {
+        let mut schema_builder = Schema::builder();
+        schema_builder.add_text_field("content", TEXT | STORED);
+        let schema = schema_builder.build();
+
+        let index = Index::create_in_ram(schema.clone());
+        let mut index_writer = index.writer(50_000_000)?;
+        let content = schema.get_field("content")?;
+
+        while let Some(document) = self.next().await {
+            index_writer.add_document(doc!(content => document.content))?;
+        }
+        index_writer.commit()?;
+
+        Ok((index, schema))
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct FrozenQuerySubset {
+    queries: Vec<String>,
+    queries_from_dataset: String,
+    dataset_queries_tested_against: String,
+    min_results_per_query: usize,
+}
+
+#[derive(Debug, Clone)]
+pub struct Document {
+    pub content: String,
+    pub metadata: HashMap<String, String>,
+}
+
+pub trait DocumentDataset
+where
+    Self: Sized,
+{
+    fn init() -> impl Future<Output = Result<Self>> + Send;
+    fn get_name(&self) -> &'static str;
+    fn create_documents_stream(
+        &self,
+    ) -> impl Future<Output = Result<impl Stream<Item = Result<Document>>>> + Send;
+}
+
+/// A query dataset is a specialized `DocumentDataset` where the documents are queries.
+pub trait QueryDataset: DocumentDataset
+where
+    Self: Sized,
+{
+    /// Returns a subset of queries from the dataset that have at least `min_results_per_query` results in the `corpus_dataset`.
+    /// The subset will contain at most `max_num_of_queries` queries.
+    ///
+    /// Because constructing this subset can be expensive (and different subsets may lead to different downstream test results), by default the constructed subset is stored in the `dataset_files/` directory in the root of this crate.
```

Comment:
we should use git lfs for that folder maybe

Reply:
These files are pretty small, assuming queries with say an average of 32 bytes 10,000 queries (more than we would ever need) is only 320 kB.